### ISSUE TITLE
API Testing: Authentication via certificate

### ISSUE THREAD

**Issue by hiobjan** [microsoft/playwright#13426]
Is it possible to use a client certificate for authentication in a REST API call?

**Comment by mxschmitt** [comment]
This is not yet possible, I'll mark this issue as a feature request. (For others: please upvote it if you are interested in that feature).

**Comment by csirwin** [comment]
Is there any status on this issue?

**Comment by mxschmitt** [comment]
No status yet! But if you update, it gets prioritized higher.

**Comment by a7md0** [comment]
Any update on this?

I think this could be implemented as the only requirement is to pass the parameters `cert` & `key` down to the HTTPS agent

1. The type `channels.APIRequestContextFetchParams` would need some adjustment to allow the two properties (declared in `packages\protocol\src\channels.ts`)
2. The type `FetchRequestOptions` need adjustment to have the default values as well(declared in `packages\playwright-core\src\server\fetch.ts`)
3. The file `packages/playwright-core/src/server/fetch.ts` would need to have few adjustments to pass these parameters under the function `fetch`.

My initial instinct was to adjust these types to have an object with the below type from node HTTPS typing, to have all TLS properties. However, I found out there's client to server messaging using RPC, which would increase the complexity of callback functions. Thus, only including the required properties `cert` & `key` would be enough to just cover this feature.
```ts
import * as tls from 'node:tls';
export type APIRequestContextFetchSecureContext = tls.SecureContextOptions & {
  checkServerIdentity?: typeof tls.checkServerIdentity | undefined;
  rejectUnauthorized?: boolean | undefined; // Defaults to true
  servername?: string | undefined; // SNI TLS Extension
}
```
```ts
export type APIRequestContextFetchParams = {
  // ...
  secureContext: APIRequestContextFetchSecureContext,
}
```

From what I noticed that types are code generated using `utils\generate_channels.js` & `packages\protocol\src\protocol.yml`, and I'm not sure if this should be modified directly, and to which type to map the new properties

**Comment by wszychta** [comment]
Hello @mxschmitt ,

We are starting to work with `playwright`, but we have discovered that lacking of this feature is totally blocking us. Can you consider adding higher priority for this issue?

**Comment by a7md0** [comment]
Any update on this? or any direction where we can contribute to implement this?

**Comment by mxschmitt** [comment]
@a7md0 I can't promise an eta but we are working on it!

**Comment by dvittor-cuscal** [comment]
I just spend the last few hours on this, only to realise it's not implemented. 
Would be awesome to have this, as all our APIs are behind MTLS certificates. 
Here is the proposed code from the client end.

We'd need to pass the parameter `agent` into the options.

```
      import https from 'https';
      import fs from 'fs';
      import path from 'path';

      const data = {};
      const headers = {};
      const timeout = 5000;
      const url = 'https://example.com/path?id=1';
      const index = url.indexOf('/', 8);
      const [baseUrl, urlPath] = [url.substring(0,index), url.substring(index)];
      const host = baseUrl.replace(/^http(s)?:\/\//i, '');
      const port = 443;
      const rejectUnauthorized = false;
      console.log( { baseUrl, host, port, urlPath });
      const certPath = [__dirname, '..', 'certs'].join(path.sep);
      const certFile = [certPath, 'cert.pem'].join(path.sep);
      const keyFile = [certPath, 'key.pem'].join(path.sep);
      const cert = fs.readFileSync(certFile, 'utf8');
      const key = fs.readFileSync(keyFile, 'utf8');
      console.log({ cert, key });
    
      const agentOptions: https.AgentOptions = {
        host,
        port,
        path: urlPath,
        cert: cert.replace(/(\r\n|\r|\n)/g, '\n'),
        key: key.replace(/(\r\n|\r|\n)/g, '\n'),
        rejectUnauthorized,
        timeout,
      };
      const agent = new https.Agent(agentOptions);
      console.log({agent});
      const options = {
        headers,
        data,
        agent,
        timeout,
      };
      
      const response = await request.post(url, options);
```

**Comment by mxschmitt** [comment]
@dvittor-cuscal we probably release this in the next version of Playwright - if you want, you can help us beta-test.

Documentation: [TestOptions](https://playwright.dev/docs/next/api/class-testoptions#test-options-client-certificates) or [newContext](https://playwright.dev/docs/next/api/class-apirequest#api-request-new-context-option-client-certificates)
Canary version: https://playwright.dev/docs/canary-releases

**Comment by dvittor-cuscal** [comment]
Hi @mxschmitt. Yeah, would be happy to help test this. Let me try the canary release, and let you know how it goes.

**Comment by dvittor-cuscal** [comment]
Note: I've found a workaround, i.e. to just use axios in place of the built in playwright request. 

I do hope in future I wont need to do this, but for now the following code is working for me:
```
      const agentOptions = getAgentOptions({ urlPath: url, timeout });
      const agent = new https.Agent(agentOptions);
      const config: AxiosRequestConfig = {
        httpsAgent: agent,
        headers,
      }

      let response;
      try {
        response = await axios.post(Env.URL+url, data, config);
      } catch (e) {
        response = (e as AxiosError).response;
      }
      const { status, statusText, data: resData } = response;

      expect(response.status).toBe(400);
      expect(resData.error_code).toBe("ERR01");
```

**Comment by mxschmitt** [comment]
This feature is implemented, and will be part of v1.46. We would appreciate, if you could help us beta test it:

- Documentation: [TestOptions](https://playwright.dev/docs/next/api/class-testoptions#test-options-client-certificates) or [newContext](https://playwright.dev/docs/next/api/class-apirequest#api-request-new-context-option-client-certificates)
- Canary version: https://playwright.dev/docs/canary-releases

**Comment by wszychta** [comment]
Hello,
good to know that this feature is available for use. I have checked provided by @mxschmitt websites, but when I change documentation to .NET then I can see that it is not available. Will this feature be available in .NET also?

**Comment by mxschmitt** [comment]
Yes, it will be available for .NET/Java/Python as well - but for them we don't have Canary versions - so you need to wait a bit until we release 1.46.
